File: app/src/main/java/com/example/miniproject/Utils/AppExecutors.java
package com.example.miniproject.Utils;

import android.os.Handler;
import android.os.Looper;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

public class AppExecutors {
    //A singleton class
    private static final int threadPoolSize = 3;
    private static final Object LOCK = new Object();
    private static AppExecutors sInstance;
    //an executor to run commands on disk
    private final Executor diskIO;
    //an executor to run commands on network
    private final Executor networkIO;
    //an executor to run commands on main thread
    private final Executor mainThreadIO;

    private AppExecutors(Executor diskIO, Executor networkIO, Executor mainThreadIO) {
        this.diskIO = diskIO;
        this.networkIO = networkIO;
        this.mainThreadIO = mainThreadIO;
    }

    public static AppExecutors getInstance() {
        if (sInstance == null) {
            synchronized (LOCK) {
                sInstance = new AppExecutors(Executors.newSingleThreadExecutor(), Executors.newFixedThreadPool(threadPoolSize),
                        new MainThreadExecutor());
            }
        }
        return sInstance;
    }

    public Executor getDiskIO() {
        return diskIO;
    }

    public Executor getNetworkIO() {
        return networkIO;
    }

    public Executor getMainThreadIO() {
        return mainThreadIO;
    }

    private static class MainThreadExecutor implements Executor {
        private Handler mainThreadHandler = new Handler(Looper.getMainLooper());

        @Override
        public void execute(Runnable command) {
            mainThreadHandler.post(command);
        }
    }

}

File: app/src/main/java/com/example/miniproject/utils/AppExecutors.kt
package com.example.miniproject.utils

import android.os.Handler
import android.os.Looper
import java.util.concurrent.Executor
import java.util.concurrent.Executors

class AppExecutors private constructor(//an executor to run commands on disk
    val diskIO: Executor
) {

    private class MainThreadExecutor : Executor {
        private val mainThreadHandler = Handler(Looper.getMainLooper())
        override fun execute(command: Runnable) {
            mainThreadHandler.post(command)
        }
    }

    companion object {
        //A singleton class
        private const val threadPoolSize = 3
        private val LOCK = Any()
        private var sInstance: AppExecutors? = null
        val instance: AppExecutors?
            get() {
                if (sInstance == null) {
                    synchronized(LOCK) {
                        sInstance = AppExecutors(
                            Executors.newSingleThreadExecutor()
                        )
                    }
                }
                return sInstance
            }
    }
}